dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java

```diff
@@ -554,18 +558,29 @@ public static String generateStorageIdentifier() {
         return storageIdentifier;
     }
 
-    public static void createIngestFailureReport(DataFile dataFile, String message) {
-        createIngestReport(dataFile, IngestReport.INGEST_STATUS_FAILURE, message);
+    public static void createIngestFailureReport(DataFile dataFile, IngestException ingestException) {
+        FileUtil.createIngestFailureReport(dataFile,
+                                           ingestException.getErrorKey(),
+                                           ingestException.getBundleArguments().toArray(new String[0]));
```

Comment:
`toArray` straight into `Arrays.asList`...

Reply:
Most callers are using single argument, I could overload the method but is it actually worth it?

Reply:
```
void createIngestFailureReport(DataFile dataFile, IngestError errorKey, String... arguments) {
    List<String> args = Optional.ofNullable(arguments).map(Arrays::asList).orElseGet(Collections::emptyList);
    return createIngestFailureReport(dataFile, errorKey, args);
}

void createIngestFailureReport(DataFile dataFile, IngestError errorKey, List<String> arguments) {
    IngestReport errorReport = new IngestReport();
    errorReport.setFailure();
    errorReport.setErrorKey(errorKey);
    errorReport.setDataFile(dataFile);
    errorReport.setReportArguments(arguments);
    dataFile.setIngestReport(errorReport);
}
```
The below 2 methods I'd replace with the two above.

Also, now that I look at it, wouldn't it make more sense to have those factory methods in `IngestReport` class? And drop coupling with `DataFile` by dropping the last line, just return `IngestReport` and let the caller use the object as desired?

Reply:
> The below 2 methods I'd replace with the two above.

I could add another with `List<String>` but I'm hesitant about removing `reateIngestFailureReport(DataFile dataFile, IngestError errorKey)`.
It would force the caller to either pass empty list every time or pass null which is kinda an anti-pattern.

> Also, now that I look at it, wouldn't it make more sense to have those factory methods in IngestReport class? And drop coupling with DataFile by dropping the last line, just return IngestReport and let the caller use the object as desired?

Makes sense, but I also have some doubts. Is making those kind of methods in Entities is correct? I always thought about it more of like an model.

Reply:
Varargs variant also covers the 2-argument call. You don't need to pass `null` as third parameter. In fact you event can't do that as it'd result in an ambiguous call error, as it would not know if by `null` you mean a `List<String>` or a varargs `String`.

As for placing factory method inside the class it's fabricating, it seems much more natural than an arbitrary placement in some utils class with very unclear responsibility. Even if said class is also an entity, factory methods are a kind of constructors, and you place constructors inside the class you're constructing.